src/plugins/vis_type_vega/public/line_vega_spec_expression.ts

```diff
@@ -0,0 +1,246 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Any modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { cloneDeep } from 'lodash';
+import { i18n } from '@osd/i18n';
+import {
+  ExpressionFunctionDefinition,
+  OpenSearchDashboardsDatatable,
+} from '../../expressions/public';
+import { VegaVisualizationDependencies } from './plugin';
+
+type Input = OpenSearchDashboardsDatatable;
+type Output = Promise<string>;
+
+interface Arguments {
+  visLayers: string | null;
+  visParams: string;
+  dimensions: string;
+}
+
+export type VegaSpecExpressionFunctionDefinition = ExpressionFunctionDefinition<
+  'vega_spec',
+  Input,
+  Arguments,
+  Output
+>;
+
+const createSpecFromDatatable = (
+  datatable: OpenSearchDashboardsDatatable,
+  visParams: string,
+  dimensionsString: string
+): object => {
+  // TODO: we can try to use VegaSpec type but it is currently very outdated, where many
+  // of the fields and sub-fields don't have other optional params that we want for customizing.
+  // For now, we make this more loosely-typed by just specifying it as a generic object.
+  const spec = {} as any;
+  const xAxis = datatable.columns[0];
+
+  const parseParams = JSON.parse(visParams);
+  const dimensions = JSON.parse(dimensionsString);
+  const legendPosition = parseParams.legendPosition;
+
+  // Get time range for the data in case there is only data for a small range so it will show the full time range
```

Comment:
just to confirm - is the problem here that the datatable may have a smaller range than the contextual time range, so you're adding time buckets at the beginning and end such that the chart will still show the correct overall time range?

Reply:
Yes, if it is not included, the data show is where we actually have data. For example if there is only data 90 days ago and nothing starting from 60 days ago, it would show the graph from 90 days ago to 60 days ago instead of the full 90 days and showing the data stopped 60 days ago.

Reply:
The logic for adding annotations is that it takes the associated event timestamp and fits it into the closest time bucket / row that's specified in the datatable. This would cause problems if the time buckets aren't equally spaced from begin timestamp to end timestamp.

I think we may need a more complex helper fn here to fill in empty rows instead of adding one row at the beginning and one row at the end.

Reply:
Looking into the agg service params, we will have to modify the logic to continue creating buckets. I will look how feasible that is to modify that as that can be risky or I will do the same as vislib and create empty buckets in the spec.

Reply:
It looks to be due to local setup issue, so will go ahead with the approach below:
```
"scale": {"domain": [1500000000000, 1500900000000]}
```